mysql/resource_grant.go

```diff
@@ -16,16 +17,225 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
-type MySQLGrant struct {
+type ObjectT string
+
+var (
+	kProcedure ObjectT = "PROCEDURE"
+	kFunction  ObjectT = "FUNCTION"
+	kTable     ObjectT = "TABLE"
+)
+
+type MySQLGrant interface {
+	GetId() string
+	SQLGrantStatement() string
+	SQLRevokeStatement() string
+	GetUserOrRole() UserOrRole
+	GrantOption() bool
+}
+
+type MySQLGrantWithDatabase interface {
+	MySQLGrant
+	GetDatabase() string
+}
+
+type MySQLGrantWithTable interface {
+	MySQLGrantWithDatabase
+	GetTable() string
+}
+
+type MySQLGrantWithPrivileges interface {
+	MySQLGrant
+	GetPrivileges() []string
+}
+
+type PrivilegesPartiallyRevocable interface {
+	SQLPartialRevokePrivilegesStatement(privilegesToRevoke []string) string
+}
+
+type UserOrRole struct {
+	Name string
+	Host string
+}
+
+func (u UserOrRole) IDString() string {
+	if u.Host == "" {
+		return u.Name
+	}
+	return fmt.Sprintf("%s@%s", u.Name, u.Host)
+}
+
+func (u UserOrRole) SQLString() string {
+	if u.Host == "" {
+		return fmt.Sprintf("'%s'", u.Name)
+	}
+	return fmt.Sprintf("'%s'@'%s'", u.Name, u.Host)
+}
+
+func (u UserOrRole) Equals(other UserOrRole) bool {
+	if u.Name != other.Name {
+		return false
+	}
+	if (u.Host == "" || u.Host == "%") && (other.Host == "" || other.Host == "%") {
+		return true
+	}
+	return u.Host == other.Host
+}
+
+type TablePrivilegeGrant struct {
 	Database   string
 	Table      string
 	Privileges []string
+	Grant      bool
+	UserOrRole UserOrRole
+	TLSOption  string
+}
+
+func (t *TablePrivilegeGrant) GetId() string {
+	return fmt.Sprintf("%s:%s", t.UserOrRole.IDString(), t.GetDatabase())
+}
+
+func (t *TablePrivilegeGrant) GetUserOrRole() UserOrRole {
+	return t.UserOrRole
+}
+
+func (t *TablePrivilegeGrant) GrantOption() bool {
+	return t.Grant
+}
+
+func (t *TablePrivilegeGrant) GetDatabase() string {
+	if strings.Compare(t.Database, "*") != 0 && !strings.HasSuffix(t.Database, "`") {
+		return fmt.Sprintf("`%s`", t.Database)
+	}
+	return t.Database
+}
+
+func (t *TablePrivilegeGrant) GetTable() string {
+	if t.Table == "*" || t.Table == "" {
+		return "*"
+	} else {
+		return fmt.Sprintf("`%s`", t.Table)
```

Comment:
You do check for backticks for t.Database and not here. Is there any reason to do it in two ways?

Reply:
After further thought, I don't think we need this anymore. The regex I wrote for capturing database and object should exclude backticks entirely, which makes this code a bunch easier to reason about:

https://github.com/petoju/terraform-provider-mysql/pull/102/files/48804a73f5cfbb1579743d10f0bc6d9d3184a9a9#diff-b410da4f476967e3ec280701a3dcf72108093885403a52fe2753d8706a6c0d72R735

Reply:
Thanks, that looks good!